src/m3ninx/index/regexp.go

```diff
@@ -0,0 +1,323 @@
+}
+
+func ensureRegexpUnanchoredHelper(parsed *syntax.Regexp, leftmost, rightmost bool) (output *syntax.Regexp, changed bool, err error) {
+	// short circuit when we know we won't make any changes to the underlying regexp.
+	if !leftmost && !rightmost {
+		return parsed, false, nil
+	}
+
+	newRegexp, changed := parsed, false
+	switch parsed.Op {
+	case syntax.OpBeginLine, syntax.OpEndLine:
+		// i.e. the flags provided to syntax.Parse did not include the `OneLine` flag, which
+		// should never happen as we're using syntax.Perl which does include it (ensured by a test
+		// in this package).
+		return nil, false, fmt.Errorf("regular expressions are forced to be single line")
+	case syntax.OpBeginText:
+		if leftmost {
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}
+			changed = true
+		}
+	case syntax.OpEndText:
+		if rightmost {
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpEmptyMatch,
+				Flags: parsed.Flags,
+			}
+			changed = true
+		}
+	case syntax.OpCapture:
+		// because golang regexp's don't allow backreferences, we don't care about maintaining capture
+		// group namings and can treate captures the same as we do conactenations.
+		fallthrough
+	case syntax.OpConcat:
+		// strip left-most '^'
+		if l := len(parsed.Sub); leftmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost && l == 1)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+		// strip right-most '$'
+		if l := len(parsed.Sub); rightmost && l > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[l-1], leftmost && l == 1, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[l-1] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpAlternate:
+		// strip left-most '^' and right-most '$' in each sub-expression
+		for idx := range parsed.Sub {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[idx], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			parsed.Sub[idx] = newRe
+			changed = changed || c
+		}
+	case syntax.OpQuest:
+		if len(parsed.Sub) > 0 {
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if c {
+				parsed.Sub[0] = newRe
+				changed = true
+			}
+		}
+	case syntax.OpStar:
+		if len(parsed.Sub) > 0 {
+			original := deepCopy(parsed)
+			newRe, c, err := ensureRegexpUnanchoredHelper(parsed.Sub[0], leftmost, rightmost)
+			if err != nil {
+				return nil, false, err
+			}
+			if !c {
+				return parsed, false, nil
+			}
+			newRegexp = &syntax.Regexp{
+				Op:    syntax.OpConcat,
+				Flags: parsed.Flags,
+				Sub: []*syntax.Regexp{
+					&syntax.Regexp{
+						Op:    syntax.OpQuest,
+						Flags: parsed.Flags,
+						Sub: []*syntax.Regexp{
+							newRe,
+						},
+					},
+					original,
```

Comment:
Hm, what's the reasoning for constructing a preceding "question"/"optional" sub-expression and concating with "star"/"any" sub-expression?  Also could the original not still contain a sub-expression with "^" or "$"?

Reply:
talked offline - have to do this to preserve the Language captured by the regular expression post transformations.

Reply:
re: original - we make the deep copy to preserve any internal "^" or "$".